A59-audit-logging.md

```diff
@@ -0,0 +1,612 @@
+will never happen twice. 
+
+If users want to audit on allow, then the `DENY` RBAC will have no audit enabled
+and the `ALLOW` RBAC with `ON_ALLOW`. So audit logging will at most happen once
+when the RPC passes through both RBACs.
+
+If users want to audit on both cases, then the `DENY` RBAC needs `ON_DENY` and
+the `ALLOW` RBAC needs `ON_DENY_AND_ALLOW`.
+
+Following is the table summarizing the combinations.
+
+|Authorization Policy  |DENY RBAC          |ALLOW RBAC           |
+|----------------------|-------------------|---------------------|
+|NONE                  |NONE               |NONE                 |
+|ON_DENY               |ON_DENY            |ON_DENY              |
+|ON_ALLOW              |NONE               |ON_ALLOW             |
+|ON_DENY_AND_ALLOW     |ON_DENY            |ON_DENY_AND_ALLOW    |
+
+Again this is a subset of what users can technically do with xDS RBACs in the way
+described in the earlier section. But we think this represents the most common
+use cases.
+
+### Built-in logger types
+
+We plan to implement the stdout logger as a built-in logger type. The type is
+named `stdout_logger`. No configuration fields are currently defined for this
+logger and it outputs log entries to stdout in JSON format.
+
+Here is the example JSON configuration in the xDS case (see [PR](https://github.com/envoyproxy/envoy/pull/26453)).
+
+```json
+{
+  "name": "user-defined-logger-name",
+  "typed_config": {
+    "@type": "envoy.extensions.rbac.audit_loggers.stream.v3.StdoutAuditLog",
+  }
+}
+```
+
+Here is the example in the authorization policy case.
+
+```json
+{
+  "name": "stdout_logger",
+}
+```
+
+Following is an example log entry.
+
+```json
+{"grpc_audit_log":{"timestamp":"2006-01-02T15:04:05.999999999Z07:00","rpc_method":"/pkg.Service/Foo","principal":"spiffe://foo/user1","policy_name":"example_policy","matched_rule":"admin_access","authorized":true}}
+```
+
+The timestamp in RFC3339 format with nanosecond-level precision represents the
+time when the logger is invoked.
+
+More types of loggers may be designed and implemented in the future. For users
+that just need to use the built-in loggers, everything will be configured in
+either the xDS RBAC filter or gRPC authorization policy. No additional code is
+required.
+
+### Language Specific APIs for third-party logger implementation
+
+Some users may want to have their own audit logging logic which built-in
+loggers do not fullfil. This section is specifically about public APIs we
+will expose for users to implement their own types of audit loggers. Below are
+two important features that apply to all the languages.
+
+As is also documented in all the languages below, the logging function is
+executed by gRPC synchronously during the authorization process. Therefore,
+its implementation should not block the RPC. When needed, it should invoke any
+long running operations asynchronously so that the function itself returns
```

Comment:
I suggest changing the wording of this sentence to make its meaning a bit clearer:

If a logger implementation needs to trigger any long-running work, it must do so asynchronously so that the function itself returns promptly.

Reply:
Suggestion taken.